File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) {
            return result;
        }
        generateCombinations(digits, 0, "", result);
        return result;
    }

    private void generateCombinations(String digits, int index, String current, List<String> result) {
        if (index == digits.length()) {
            result.add(current);
            return;
        }

        String letters = getLetters(digits.charAt(index));
        for (char letter : letters.toCharArray()) {
            generateCombinations(digits, index + 1, current + letter, result);
        }
    }

    private String getLetters(char digit) {
        switch (digit) {
            case '2':
                return "abc";
            case '3':
                return "def";
            case '4':
                return "ghi";
            case '5':
                return "jkl";
            case '6':
                return "mno";
            case '7':
                return "pqrs";
            case '8':
                return "tuv";
            case '9':
                return "wxyz";
            default:
                return "";
        }
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits):
        result = []
        if digits is None or len(digits) == 0:
            return result
        self.generateCombinations(digits, 0, "", result)
        return result

    def generateCombinations(self, digits, index, current, result):
        if index == len(digits):
            result.append(current)
            return
        letters = self.getLetters(digits[index])
        for letter in letters:
            self.generateCombinations(digits, index + 1, current + letter, result)

    def getLetters(self, digit):
        digit_to_letters = {
            '2': "abc",
            '3': "def",
            '4': "ghi",
            '5': "jkl",
            '6': "mno",
            '7': "pqrs",
            '8': "tuv",
            '9': "wxyz"
        }
        return digit_to_letters.get(digit, "")

